### PR TITLE
CORE-12440: Fix avro field names in C++

### PR DESCRIPTION
**Pull request by nguyen-andrew**
This PR adds proper simple name validation in alignment with the [Avro spec](https://avro.apache.org/docs/1.12.0/specification/#names) in the C++ codebase. It introduces NameValidationTests, replaces the old character checks with validateSimpleName, and updates record, enum, and JSON compilation paths to correctly reject invalid names.

Related: https://issues.apache.org/jira/browse/AVRO-4155

### CODE REVIEW

File: lang/c++/test/NameValidationTests.cc [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    "Invalid Name", // contains space
+    "Invalid-Name", // contains hyphen
+    "com.example.1Invalid", // invalid full name
+    "" // empty name
+};
+
+void testValidSchemaNames() {
+    for (const auto& name : validSchemaNames) {
+        BOOST_TEST_MESSAGE("Testing valid schema name: " << name);
+        BOOST_CHECK_NO_THROW((avro::RecordSchema{name}));
```
Comment (pgellert) [inline lang/c++/test/NameValidationTests.cc]: nit: why the double `((...))` here?

File: lang/c++/test/NameValidationTests.cc [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "avro/Schema.hh"
+
+#include <boost/test/included/unit_test.hpp>
+#include <boost/test/unit_test.hpp>
+
+std::vector<std::string> validSimpleNames = {
```
Comment (pgellert) [inline lang/c++/test/NameValidationTests.cc]: Make these `const` to prevent modification, and `static` to give them internal linkage so they aren't visible outside this file.

File: lang/c++/impl/Node.cc [hunk 3]
```diff
@@ -105,18 +105,30 @@ static bool invalidChar1(char c) {
     return !isalnum(c) && c != '_' && c != '.' && c != '$';
 }
 
-static bool invalidChar2(char c) {
-    return !isalnum(c) && c != '_';
+void validateSimpleName(const std::string &name) {
+    if (name.empty()) {
+        throw Exception("Invalid name: " + name);
```
Comment (pgellert) [inline lang/c++/impl/Node.cc]: nit: I'd maybe rephase this to "Name cannot be empty" to make the cause of the exception more apparent

File: lang/c++/impl/Node.cc [hunk 4]
```diff
@@ -105,18 +105,30 @@ static bool invalidChar1(char c) {
...
+    }
+    
+    char first = name[0];
+    if (!(std::isalpha(first) || first == '_')) {
```
Comment (pgellert) [inline lang/c++/impl/Node.cc]: `std::isalpha` and the other `<cctype>` functions take an int, but they expect the value to be either EOF or something representable as unsigned char. If `first` is signed and you pass a negative value (for example a byte >127), the behavior is undefined. To avoid this, you need to cast the byte to unsigned char first so the promoted value is in the 0–255 range.

From https://en.cppreference.com/w/cpp/string/byte/isalpha.html:
> The behavior is undefined if the value of ch is not representable as unsigned char and is not equal to [EOF](https://en.cppreference.com/w/cpp/io/c.html).

These are also locale-dependent methods, so to avoid that complexity, it's probably simplest to just define two small local helper functions for this.

Comment (pgellert) [reply lang/c++/impl/Node.cc]: It's probably worth adding a unit test with these negative chars as well. E.g.; `"éName"`

File: lang/c++/test/NameValidationTests.cc [hunk 5]
```diff
@@ -104,18 +106,140 @@ void testInvalidEnumSymbols() {
...
     ts->add(BOOST_TEST_CASE(&testValidSchemaNames));
     ts->add(BOOST_TEST_CASE(&testInvalidSchemaNames));
-    
     ts->add(BOOST_TEST_CASE(&testValidFieldNames));
     ts->add(BOOST_TEST_CASE(&testInvalidFieldNames));
-
     ts->add(BOOST_TEST_CASE(&testValidEnumSymbols));
     ts->add(BOOST_TEST_CASE(&testInvalidEnumSymbols));
+    
+    // JSON compilation validation tests
```
Comment (pgellert) [inline lang/c++/test/NameValidationTests.cc:235]: Thanks for the nice commit-by-commit changes 👏

File: lang/c++/impl/Compiler.cc [hunk 6]
```diff
@@ -293,6 +293,7 @@ static void getCustomAttributes(const Object &m, CustomAttributes &customAttribu
 static Field makeField(const Entity &e, SymbolTable &st, const string &ns) {
     const Object &m = e.objectValue();
     string n = getStringField(e, m, "name");
+    validateSimpleName(n);
     vector<string> aliases;
     string aliasesName = "aliases";
     if (containsField(m, aliasesName)) {
```
Comment (pgellert) [inline lang/c++/impl/Compiler.cc:299]: We don't need to validate aliases because of the spec, right?
> Aliases should be valid names, but this is not required: any string is accepted as an alias. When aliases are used “to map a writer’s schema to the reader’s” (see above), this allows schema evolution to correct illegal names in old schemata.

**Review by pgellert — COMMENTED** [review]
Looks good.

This is a breaking change, so we should release this on a major in redpanda and add a release note.

Let's discuss if this will need some data rewriting now in redpanda SR. If we have already serialized field names like the ones in the CORE-12440 ticket, we might need to rewrite those into a valid format now.

File: lang/c++/impl/Node.cc (hunk 4 above)
Comment (nguyen-andrew) [reply lang/c++/impl/Node.cc]: Good catch, updated this + added additional test to validate that non-ascii characters are rejected.

File: lang/c++/impl/Compiler.cc (hunk 6 above)
Comment (nguyen-andrew) [reply lang/c++/impl/Compiler.cc:299]: That's what I was thinking as well.

**Comment by nguyen-andrew** [comment]
[Force push](https://github.com/redpanda-data/avro/compare/20447314ada50e37346b0ba7824266d974c71dd1..276bd26ec86c9ca5ddee4dc2f47ad3b418101c94) to address comments - most were minor tweaks, but the important one was updating the name check to use simple, locale-independent ASCII check functions instead of cctype functions. Also added tests to verify that non-ASCII names are rejected.

File: lang/c++/test/NameValidationTests.cc (hunk 1 above)
Comment (nguyen-andrew) [reply lang/c++/test/NameValidationTests.cc]: Whoops, fixed.

**Review by pgellert — APPROVED** [review]
